web/src/main/java/org/springframework/security/web/method/annotation/AuthenticationPrincipalArgumentResolver.java

```diff
@@ -173,9 +179,22 @@ public void setTemplateDefaults(AnnotationTemplateExpressionDefaults templateDef
 	 * @param parameter the {@link MethodParameter} to search for an {@link Annotation}
 	 * @return the {@link Annotation} that was found or null.
 	 */
-	@SuppressWarnings("unchecked")
-	private <T extends Annotation> T findMethodAnnotation(MethodParameter parameter) {
-		return (T) this.scanner.scan(parameter.getParameter());
+	private AuthenticationPrincipal findMethodAnnotation(MethodParameter parameter) {
+		if (this.useAnnotationTemplate) {
+			return this.scanner.scan(parameter.getParameter());
+		}
+		AuthenticationPrincipal annotation = parameter.getParameterAnnotation(this.annotationType);
```

Comment:
This algorithm does not support `aliasFor`. Since that was also added in 6.4 and is unrelated to the template expressions, let's please change everything from line 186 to 197 to the following:

```java
return MergedAnnotations.from(parameter.getParameter(), MergedAnnotations.SearchStrategy.INHERITED_ANNOTATIONS, RepeatableContainers.none())
				.get(annotationType).synthesize(MergedAnnotation::isPresent).orElse(null);
```

Reply:
But If do like that, parameter annotation can't resolve from interface or parent that is #16177 mentioned.By research `MergedAnnotations `source code `MergedAnnotations.SearchStrategy` only effect class or method .If the `AnnotatedElement` is parameter , only the annotations on the current element will not include inheritance.

Reply:
Please take a look at this and let me know what you think: https://github.com/spring-projects/spring-security/commit/b8bfbace5f993fc452ebe55e2eeef5a7a7d911c8

Reply:
Sorry, I misinterpreted you.